Simplify derivatives in pysindy.py

Jacob-Stevens-Haas opened the pull request:
Removes a lot of nested conditionals and ~150 SLOC.  All derivatives/integrals are calculated in a single line of code inside pysindy.py. Passes all tests locally, but I'm leaving it as a draft until I can check that the notebooks all run.

As the former version of pysindy/pysindy.py said:
```python
 # Note: This block for calculating x_dot for weak and pde libs is identical
 # to the one in score. Should we redefine the differentiation_method to cover
 # the weak and pde library cases more generally?
 # The validate_input should probably explicity handle the weak and pde cases...
```
This PR addresses that concern with two abstractions:
1. Feature libraries are responsible for addressing how differentiation occurs via their `calc_trajectory()` method.  For `BaseFeatureLibrary`, that simply means applying the differentiation function passed.  For `PDELibrary`, that means ignoring the differentiation library and using FiniteDifference.  For `WeakPDELibrary`, that means ignoring the differentiation library passed and using `convert_udot_integral`.  `GeneralizedLibrary` chooses which method depending on which types of library it has, in an existing loop.  These are how `SINDy.fit()` (and other methods) was previously working, but it allows all of the conditionals to be replaced with: `self.feature_library.calc_trajectory(self.differentiation_method, x, t)`.  I chose that name because it's called in an iterator within `_process_multiple_trajectories`, but happy to change it.
2. Feature libraries are responsible for addressing how input validation occurs.  For `BaseFeatureLibrary`, that simply means applying `utils.validate_input()`.  For `PDELibrary` and `WeakPDELibrary` (and when applicable, `GeneralizedLibrary`) just return `x`.

It also adds two more abstractions that assist code deletion and DRY
3. `_adapt_to_multiple_trajectories()` makes all trajectories lists, so code handling single trajectories is removed
4.  `_zip_like_sequence()` allows combining the many conditionals around whether t is a sequence or a scalar:
```python
if isinstance(t, Sequence):
    [something(xi, ti) for xi, ti in zip(x, t)
else:
    [something(xi, t) for xi in x)
```
becomes
```python
[something(xi, ti) for xi, ti in _zip_like_sequence(x, t)]
```

Only immediately obvious con of this PR is a slight performance hit in the non-PDE cases.  The PDE cases previously ran `validate_input()` *after* their differentiation/integration step.  Now non-PDE cases do so as well before differentiation.  Using the current form of `calc_trajectory()` isn't compatible with this method unless `BaseFeatureLibrary.calc_trajectory()` would return `x, x_dot` instead of just `x_dot`.  I'd love for that to be the case (see #182 ), but it was a design decision I wanted to kick back to the group in case objections.

Other issues:
* There's still some [weird conditional code in](https://github.com/dynamicslab/pysindy/blob/04a8374b8a579f98e26f639444316f77bdc78bed/pysindy/pysindy.py#L982) `_process_multiple_trajectories()` for when `x_dot` is explicitly given as an argument.  It assumes to handle some `WeakPDELibrary` cases when `x_dot` is given in a **different** form than if calculated from the data.  I'm assuming that test case is a mistake?
* Tests take a long time.  In many cases (esp PDE/WeakPDE cases) the test generates 100x100 matrices for data.  Maybe it's possible to reduce that?
* The feature libraries may be the right place for `validate_input`, but maybe there's a more elegant solution?

akaptano commented:
@Jacob-Stevens-Haas Great job with this. This is no longer my day job but I should be able to give this a solid look this long weekend!

Jacob-Stevens-Haas commented:
Rebased onto `master`
@znicolaou ready for you to adapt `PDELibrary` & `WeakPDELibrary` `fit()` and `tranform()` to new internal array format, keeping data in its native shape.

(WIP comment... saving progress after I lost it all earlier)

znicolaou commented:
Cool, going to take a look at this and #195 tomorrow. I'm not sure how big of a change the fit and transform will be, but hopefully it will not be too complicated since they already had to reshape the data in the beginning. Curious to see how the multiple_trajectories are going to be handled--the previous approach was convoluted.

Jacob-Stevens-Haas commented:
@znicolaou Thanks for the quick work!  I'm happy to sync branches and I'm looking at the changes you made.  Before I merge, I have a few questions/comments and I want to link to specific line numbers, though.

pysindy/pysindy.py
1. `# Deprecation warnings are ignored by default. ` - good point, so should we make it `UserWarning`?
2. `# backwards compatibility for ensemble options` - thank you for this +1
3. I have a mild preference for [inlining the call to `SampleConcatter()` ](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/pysindy.py#L429) since each step of the pipeline is already on its own line.
4. Predict logic - Thanks for catching this!  The code you commented out isn't necessary, can delete.
5. `vstacking` in `score()` - I meant to have this inside the `concat_sample_axis` function

pysindy/utils/axes.py
1. Is there a reason for [`__init__(): pass`](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/utils/axes.py#L184)?  I'd prefer to delete it since it doesn't do anything.
2. Thanks for fixing `fit()`!

pysindy/optimizers/base.py
1. Why does BaseOptimizer [downcast an `AxesArray` to `np.asarray()`](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/optimizers/base.py#L178)?
2. Thank you for fixing the relationship re: `fit()` and `_reduce()`!
3. Line 349 - good catch!
4. Line [351](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/optimizers/base.py#L351) - Good catch!  BTW, can we cast `y` as `AxesArray` and change `y.shape[1]` to `y.shape[y.ax_coord]`?
5. Line [370](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/optimizers/base.py#L370) I think the old way had one subtle advantage - if `x` and `y` have axes that are not aligned but still have the same number of samples, the old way still selects the correct axis from `x_dot`

pysindy/optimizers/sindy_optimizer.py
1. These changes both reflect my mistake in forgetting the `np.concatenate` step of `concat_sample_axis()` (as mentioned in pysindy/pysindy.py item 5).  Once I fix that, these changes shouldn't be necessary.

pysindy/feature_library/polynomial_library.py
1. All good, thanks!

Jacob-Stevens-Haas commented:
Ooh, looks like there might be an issue if feature libraries are passed a list of sparse arrays

Also, in pysindy/pysindy.py Line [490](https://github.com/dynamicslab/pysindy/blob/787f4d1f5eb9f6212bb04ef3dc7f0bda2e07e27b/pysindy/pysindy.py#L490) changes the time axis to sample for `x`, but then `u` should also be changed.

Jacob-Stevens-Haas commented:
> One pending challenge will be that we will want the user to be able to use the feature libraries alone outside of the pysindy model. This allows several hacks for more advanced usage, and is detailed in some of the notebooks. So we'll either want to expose the users to AxesArrays, or, more ideally, automatically convert input to AxesArrays if it is provided as numpy arrays in a backward-compatible manner.

Yeah, agree completely.  Question is whether its acceptable for libraries to return an AxesArray, or downcast to numpy ndarray?

Jacob-Stevens-Haas commented:
FWIW, the last commit to pass all tests was https://github.com/dynamicslab/pysindy/pull/185/commits/d4f9a18c5d64fcd129574a966e1eab97f303a6f7.  Currently at around 428/578 pass.

znicolaou commented:
Things are coming along! I will work on updating all the feature libraries today.  In response to your points above:

pysindy/pysindy.py

1. inlining is good!
2. yeah, I was curious about whether this should stack the samples! It will simplify a lot of my changes from last time.

pysindy/utils/axes.py

1. No real reason, just thought I might have to add something later. But probably not.

pysindy/optimizers/base.py

1. I kept the np.asarray that was there previously just to make sure I didn't break anything. But I'd try to preserve the AxesArray as well if possible.
2. I think casting y to AxesArray is a good idea.
3. At some point, I ran this where x_dot had not been cast to AxesArray with comprehended axes, so an error was raised. If we ensure it will work, I'd also use x_dot.

About the sparse arrays, yeah, we might have to think about that. I think the only implementation of sparse arrays is in PolynomialLibrary, and I have some doubt that it has been updated or is being used. If we could create a SparseAxesArray in the future to enable sparse support for all libraries, that would be great.

For the libraries return, I think i'd stick to AxesArray for now. Probably people will just treat them like ndarrays, which should work fine since they inherit. I was also thinking, I will try to keep the reshaping/adapting to multiple trajectories out of the library fit and transform functions, to keep them simple. I saw you pushed a decorator, which seems like good idea. But it seemed strange to repeat all the adapting/reshaping in each fit and transform, since it is already done in the model fit. (If it can all be handled by such decorators, we could also remove the redudant code from pysindy.py). I think I'd prefer users have to send lists of AxesArrays to fit and transform if they need to access nuts and bolts themselves, but open to suggestions!

znicolaou commented:
I updated all the ode libraries to take the list input (except sindy_pi, which is being replaced in Alan's weak_form_sindypi branch). I also merged with your last commit, and pushed. Hopefully will be able to finish with the pde libraries tomorrow!

Jacob-Stevens-Haas commented:
>     3. At some point, I ran this where x_dot had not been cast to AxesArray with comprehended axes, so an error was raised. If we ensure it will work, I'd also use x_dot.

I added it to my "once everything passes tests" list :laughing: 


> About the sparse arrays, yeah, we might have to think about that. I think the only implementation of sparse arrays is in PolynomialLibrary, and I have some doubt that it has been updated or is being used. If we could create a SparseAxesArray in the future to enable sparse support for all libraries, that would be great.

I "implemented" it with a `wraps_axes` decorator factory, that basically sets new attributes for an existing sparse array.  That's easier than a true multiple inheritance situation, which could probably have some serious MRO conflicts. 

 
> For the libraries return, I think i'd stick to AxesArray for now. Probably people will just treat them like ndarrays, which should work fine since they inherit. I was also thinking, I will try to keep the reshaping/adapting to multiple trajectories out of the library fit and transform functions, to keep them simple. I saw you pushed a decorator, which seems like good idea. But it seemed strange to repeat all the adapting/reshaping in each fit and transform, since it is already done in the model fit. (If it can all be handled by such decorators, we could also remove the redudant code from pysindy.py). I think I'd prefer users have to send lists of AxesArrays to fit and transform if they need to access nuts and bolts themselves, but open to suggestions!
I agree.  Currently `SINDy.predict()` returns an AxesArray regardless of input type.  Up until today, there was some odd behavior for AxesArray that users wouldn't necessarily expect with an array (e.g. `arr.all()` returned `AxesArray([True])` instead of `True`).  I've fixed that bug, but for the moment feature libraries return an ndarray when passed an ndarray, a sparse array when a sparse array is passed, and a list of AxesArrays when a list of AxesArrays is passed.  It's easy to change, however.

Jacob-Stevens-Haas commented:
Uh oh, looks like your branch also includes the commits from #195.  Is there a chance you can cherry-pick the relevant ones?

Also, [most recent commit](https://github.com/dynamicslab/pysindy/pull/185/commits/9dce5143b0457818642361f4790e2a1293865a6d) passes all Polynomial Library tests.

znicolaou commented:
I verified that the commits to #195 only effect the pde_library and weak_pde_library features (with a few inconsequential changes in pysindy.py). I went ahead and merged them into this branch now so I don't have to do a complicated merge after updating them for this pr. It should be safe to pull them, but let me know if you notice any inappropriate changes.

Sounds good! Take a look at the example 1 notebook--I think the EnsembleOptimizer was at least partly functional. There should also be some simplifications where we can remove code for the old ensembling, but I haven't looked in detail.

Jacob-Stevens-Haas commented:
Makes sense... but I'm just really struggling to understand the diffs and blame.  I'm not used to that workflow so it's hard to know what tests should be passing in a merge.  Might be better to focus on finishing the other PR in the interim while I get EnsembleOptimizer to work for the PolynomialLibrary.  At the very least until I can get a version of simplify-fitting that only fails the WeakPDELibrary, PDELibrary, and SINDyPILibrary tests, at which point it might make sense to cross the streams.

Unrelated - in the backwards compatibility changes to pysindy/pysindy.py you wrap `EnsembleOptimizer` in a `SINDyOptimizer`, which "Enables single target regressors (i.e. those whose predictions are 1-dimensional) to perform multi target regression".  Is it  the case that _all_ optimizers are designed for single targets and should be wrapped in a SINDyOptimizer when dealing with multiple targets?

znicolaou commented:
I finished updating the to-be-deprecated sindypi library, and cleaned up my test comments. I think the most recent push should be easier to navigate. Some of the old tests used the data_pde_library to fit ODE data, which is no longer appropriate, so I commented out those for now. We should probably replace them.  It looks like the only serious test failures have to do with the ensembling -- the number of models being returned differs now from the old version in some cases.  We still have some cleaning up to do, but things are looking good!

Jacob-Stevens-Haas commented:
@znicolaou  how should 
```
SINDy(library=WeakPDELibrary()).fit(u, u_dot)
```
behave?  There are some tests that only passed because the implicit reshaping didn't cause an error in `master` and the data for the tests (`data_1d_random_pde`)was randomly generated with `n_spatial == n_time == WeakPDELibrary.K`.

As far as I can tell, there's no point to passing u_dot to `SINDy.fit()` when using the WeakPDELibrary, because the later is only meaningful if the integration domains are the same.

Other than this, all the tests are passing now :)

znicolaou commented:
Hey Jake, pretty sure that's a relic of when the weak form was part of the pdelibrary. Previously, the user had to calculate and pass the weak form of udot by hand, using a utility function. We should remove or replace those tests

znicolaou commented:
@Jacob-Stevens-Haas I did a little organizing to resolve some issues that were bugging me. I moved the drop_nan_rows check outside of pysindy.py, to the sindy_optimizer instead. Previously, since the shape of x and x_dot differed in the weak form, lots of tests were required in case of dropping endpoints in derivatives. But we can simplify this a lot by deferring this check until the library.transform function has been accomplished. This allows us to remove all the checks for the weak or generalized libraries in pysindy.py, and insteady just drop the samples with nans just before the optimizer step. I pushed to the simplify-fitting branch here - take look to see if you can merge with your branch.

znicolaou commented:
@Jacob-Stevens-Haas I fixed one more bug with the ensembling (that I may have introduced in a previous commit). The EnsembleOptimizer should not set normalize_columns to be opt.normalize_columns, since that will cause it to multiply the columns by the normalization factor twice.

Jacob-Stevens-Haas commented:
@znicolaou why the SINDyOptimizer?  Is it not required in any other optimizer?

Jacob-Stevens-Haas commented:
Commenting out `import sympy as sp` in notebook 9 raises an error in cell 8
```python
sym_features = [sp.symbols(feature) for feature in features]
NameError: name 'sp' is not defined
```

Jacob-Stevens-Haas commented:
## Notebook report
1 - passes
2 - passes
3 - passes
4 - passes
5 - kernel crash
6 - fails on cell 4, score() -> predict() -> SINDyOptimizer.predict()
  ValueError: Found array with dim 3. Estimator expected <= 2.
7 - kernel crash
8 - fails in cell 9. printing w/max_iter = 2 raises ZeroDivisionError in 
  TrappingOptimizer._objective() on L 482
9 - passes (w/ import sympy as sp)
10 - kernel crash cell 14
11 - passes
12 - passes
13 - passes (removed a cell you added with a reference to x_test_sims above the cell it was defined)
14 - passes
15 - fails cell 22 with ValueError in FiniteDifference._differentiate in SINDyPILibrary
  could not broadcast input array from shape (3998,1) into shape (4000,1)

znicolaou commented:
For the SINDyOptimizer point, this optimizer wraps all others in pysindy.py, mainly to enable to unbiasing option as far as I can tell. We need to drop the nan rows from both y=x_dot and x before running the optimizers, but since the Pipeline steps will not modify y, I don't see how to add it as a Pipeline step. But we can do it in the SINDyOptimizer fit, and this will effect all other optimizers in turn.

Sorry for unclear changes, like the sympi comment. I just didn't have sympy in my environment, so commented to run and forgot to uncomment. I can't track the notebook diffs either, and just running them changes things, even if there is no change in behavior. So I've just been going cell by cell to see if the desired results is achieved. There's still places that need work. And I do need to make some tests for clarity. Lastly, I commented in your pull request about the _MultiTargetLinearRegressor -- I don't think it is being used, but we should probably double check.

FYI, The change in ensembling has effected a few things--previously, rows were dropped according to their time value, so num_samples was a fraction of len(t). But now, for the PDE libraries, we need to drop a fraction of the total samples np.product(x.shape[:-1]) instead. I made some changes in notebook 1 to drop the right number of samples, but we probably need to check everywhere for that.

Lastly, notebook 10 is running on my laptop, but takes a lot of time and memory.

Jacob-Stevens-Haas commented:
# Notebook report
----------------------------------------
1 - passes
2 - passes
3 - passes
4 - passes
5 - kernel crash
6 - fails on cell 4, score() -> predict() -> SINDyOptimizer.predict()
  ValueError: Found array with dim 3. Estimator expected <= 2.
7 - kernel crash
8 - passes
9 - passes
10 - kernel crash cell 14
11 - passes
12 - passes
13 - passes
14 - passes
15 - passes
----------------------------------

working on notebook 6 now.

Jacob-Stevens-Haas commented:
<details>
<summary>New failure in NB 7, cell 38:</summary>

```
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_21830/4079072432.py in <module>
     50     cmap="coolwarm",
     51     edgecolors="k",
---> 52     linewidths=1,
     53 )
     54 fs = 22

~/github/pysindy/env/lib/python3.7/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1410     def inner(ax, *args, data=None, **kwargs):
   1411         if data is None:
-> 1412             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1413 
   1414         bound = new_sig.bind(ax, *args, **kwargs)

~/github/pysindy/env/lib/python3.7/site-packages/matplotlib/axes/_axes.py in pcolor(self, shading, alpha, norm, cmap, vmin, vmax, *args, **kwargs)
   5782         shading = shading.lower()
   5783         X, Y, C, shading = self._pcolorargs('pcolor', *args, shading=shading,
-> 5784                                             kwargs=kwargs)
   5785         Ny, Nx = X.shape
   5786 

~/github/pysindy/env/lib/python3.7/site-packages/matplotlib/axes/_axes.py in _pcolorargs(self, funcname, shading, *args, **kwargs)
   5568                 raise TypeError('Dimensions of C %s are incompatible with'
   5569                                 ' X (%d) and/or Y (%d); see help(%s)' % (
-> 5570                                     C.shape, Nx, Ny, funcname))
   5571         else:    # ['nearest', 'gouraud']:
   5572             if (Nx, Ny) != (ncols, nrows):

TypeError: Dimensions of C (10, 50) are incompatible with X (50) and/or Y (11); see help(pcolor)
```
</details>

znicolaou commented:
I have a few pending tasks here--hopefully I can get to them this week.

1. For #195, I did merged the most recent commit before making changes to the libraries, so feel free to close it. We still need to test the sindypi functionality on a pde example. If you started anything in the notebook, please push it first!
2. I think we still have some corrections to make for the reshaping. Specifically, the user should be able to call model.fit with AxesArray input, where the axes are not in the default order. We should make new tests for this, and debug it.
3. I think it should be possible to update the pde libraries to accept lists of spatial/spatiotemporal grids, which it will apply to the list of trajectories in sequence. This would be an enhancement, so maybe better to save it for a future new pull request.
4. Lastly, we know the drop_nan functionality isn't tested for the score and predict, so we should make a new test.

Jacob-Stevens-Haas commented:
@znicolaou in addition to 3, I think 2 is an added feature.  Yes it makes sense, but it doesn't specifically conflict with this I'd prefer to add an issue and reserve that for a later PR.  I need to get going on my general exam, which means the changes I was hoping to get to when I realized that this refactoring was in order.

WRT 4, do we need the integration test "`score()` with NaNs" or just the unit test for drop_nan?  I would think the latter

I think the bigger question, @akaptano , is whether or not this should be a 2.0 release because it changes (a) the shape of returns from derivative, (b) how ensembling works when ensembling both data and the library.

PS - I finally got VSCode remote development to work on Doppio.  The notebooks and tests run a bit easier on 20 cores.  LMK if you want help setting that up.

Jacob-Stevens-Haas commented:
Alright, I pushed the fix to NB7 - running now start to finish now, but barring any errors, that should be all. This seemed to be a bug where code =/= outputs even in `master`.  But, I didn't commit the changes to metadata and outputs, just the code cells.  The LaTeX renderer is different in the jupyter running in my env, so some cells have LaTeX errors.  I also didn't change r, since I don't want to be tweaking what the notebook is trying to accomplish on this branch, just getting it to run without errors.  

@znicolaou , added test, verified it failed on `master` and passed on `simplify-fitting`, pushed.  Also, what were you talking about with NB 9?  It ran on my PC fine as-is with `sympy` installed.

This is all tests & notebooks passing, so if CI is green I think we're ready to merge.  Thus, before we merge, do we make a `1.x-master` with the idea that changing derivative shapes and how ensembling is done is worth a new major version, but we still want to support 1.x?  Or do we just merge to master because these changes are small enough and reasonably expected?

znicolaou commented:
@Jacob-Stevens-Haas @akaptano I finished up one last push tonight, completing a first draft of the notebook 9 sindypi mkdv example for both the PDELibrary and the WeakPDELibrary. I added one more argument to both libraries, `multiindices`, which allows users to override the default derivative terms included in the library (which are determined by `derivative_order`). This is helpful here because we want the maximum order of the time derivative to be 1, but the maximum order of the spatial derivatives needs to be 3 (including all the mixed order derivatives leads to too many library terms and a poor fit.) 

Presentation and documentation could be improved, but it looks like the results are comparable to those in Kardierdan's paper, so I am ready to review also.  @Jacob-Stevens-Haas, do you mind merging with your branch to put this all in this pull request? This will allow us to close #195 at the same time.

In the process, I discovered model.score was still having issues with drop_nan_rows for pde data, which I fixed with some simple changes. I also added a new test `test_score_pde` to test_pysindy.py.

Jacob-Stevens-Haas commented:
All pushes added.

akaptano commented:
Is this ready for me to a final look and a new release? I'm closing #195 and let's merge in #203 first.

akaptano commented:
Hi @Jacob-Stevens-Haas and @znicolaou, 

I am realizing I will be very busy at a conference for the next two weeks, and I don't want to delay this too much. If one of you is willing to do some last checks and write the new release message (so we don't have to wait), I would be appreciative. I can look for minor bugs and loose ends in a few weeks, and can always release a minor PR with any fixes. Let me know if you want any advice/help with the new release.

znicolaou commented:
Okay, thanks for the heads up. I should be able to contribute some to final checks and release notes this this next week. @Jacob-Stevens-Haas -- if you're available, happy to bounce some drafts and aim to merge with master when we're satisfied.

Jacob-Stevens-Haas commented on pysindy/feature_library/weak_pde_library.py:
```diff
@@ -804,40 +831,35 @@ def transform(self, x):
-        for trajectory_ind in range(self.num_trajectories):
-            x = np.reshape(x_full[trajectory_ind], (n_samples_original, n_features))
-            xp = np.empty((n_samples, self.n_output_features_), dtype=x.dtype)
+        # x = check_array(x)
+        #
+        # n_samples_original_full, n_features = x.shape
+        # n_samples_original = n_samples_original_full // self.num_trajectories
+        #
+        # if float(__version__[:3]) >= 1.0:
+        #     if n_features != self.n_features_in_:
+        #         raise ValueError("x shape does not match training shape")
+        # else:
+        #     if n_features != self.n_input_features_:
+        #         raise ValueError("x shape does not match training shape")
+        #
+        # if self.spatiotemporal_grid is not None:
+        #     n_samples = self.K
+        #     n_samples_full = self.K * self.num_trajectories
+        #
+        # xp_full = np.empty(
+        #     (self.num_trajectories, n_samples, self.n_output_features_), dtype=x.dtype
+        # )
+        # x_full = np.reshape(
+        #     x, np.concatenate([[self.num_trajectories], self.grid_dims, [n_features]])
+        # )
+        xp_full = []
```
Remove comments.

Jacob-Stevens-Haas commented on pysindy/feature_library/weak_pde_library.py:
```diff
@@ -846,22 +868,22 @@ def transform(self, x):
-            dims = np.array(x_shaped.shape)
-            dims[-1] = n_library_terms
-            funcs = np.zeros(dims)
+            # x_shaped = np.reshape(
+            #     x,
+            #     np.concatenate([self.spatiotemporal_grid.shape[:-1], [x.shape[-1]]]),
+            # )
+            # dims = np.array(x_shaped.shape)
+            # dims[-1] = n_library_terms
```
remove comments.

Jacob-Stevens-Haas commented on pysindy/feature_library/pde_library.py:
```diff
@@ -130,38 +142,87 @@ def __init__(
+            raise ValueError(
+                "temporal_grid parameter must be specified if implicit_terms "
+                " = True (i.e. if you are using SINDy-PI for PDEs)."
+            )
+        elif not implicit_terms and temporal_grid is not None:
+            raise ValueError(
+                "temporal_grid parameter is specified only if implicit_terms "
+                " = True (i.e. if you are using SINDy-PI for PDEs)."
+            )
+        if spatial_grid is not None and spatial_grid.ndim == 1:
+            spatial_grid = spatial_grid[:, np.newaxis]
+
+        if temporal_grid is not None and temporal_grid.ndim != 1:
+            raise ValueError("temporal_grid parameter must be 1D numpy array.")
+        if temporal_grid is not None or spatial_grid is not None:
+            if spatial_grid is None:
+                spatiotemporal_grid = temporal_grid
+                spatial_grid = np.array([])
+            elif temporal_grid is None:
+                spatiotemporal_grid = spatial_grid[
+                    ..., np.newaxis, :
+                ]  # append a fake time axis
+            else:
+                spatiotemporal_grid = np.zeros(
+                    (
+                        *spatial_grid.shape[:-1],
+                        len(temporal_grid),
+                        spatial_grid.shape[-1] + 1,
+                    )
+                )
+                for ax in range(spatial_grid.ndim - 1):
+                    spatiotemporal_grid[..., ax] = spatial_grid[..., ax][
+                        ..., np.newaxis
+                    ]
+                spatiotemporal_grid[..., -1] = temporal_grid
+        else:
+            spatiotemporal_grid = np.array([])
             spatial_grid = np.array([])
 
+        self.spatial_grid = spatial_grid
```
Is there a simpler way to do this?

Jacob-Stevens-Haas commented on pysindy/feature_library/pde_library.py:
```diff
@@ -107,6 +116,8 @@ def __init__(
         ensemble_indices=[0],
         periodic=False,
+        implicit_terms=False,
+        multiindices=None,
```
@znicolaou , what are the use cases for someone to pass `multiindices` directly?

Jacob-Stevens-Haas commented on pysindy/feature_library/pde_library.py:
```diff
@@ -304,68 +372,42 @@ def transform(self, x):
-                    x, np.concatenate([self.grid_dims, [num_time], [n_features]])
-                )
-                for axis in range(self.grid_ndim):
+                derivs = x.copy()
```
Can we remove `copy`?

Jacob-Stevens-Haas replied:
@znicolaou  I removed it and all tests passed... If you know a reason for the copy, lmk.

znicolaou replied:
I think this change is going to mess up mixed derivatives. I don't think the examples in the notebooks and test contain any fits that require terms like u_{xy}, but I don't think the library terms corresponding to these will be calculated correctly in the current code (I think it will only return the derivative of the last nonzero element of the multiindex). We want to loop over all axes, and accumulate the derivatives for each axis to calculate the full derivative, so the old code first set `derivs=x.copy()`, the calculated `derivs=FiniteDifference(derivs)`, so that each time it loops over an axis, it accumulates a new derivative. My memory is that if you don't include the x.copy(), x itself gets modified during the loop since the assignment is by reference, leading to incorrect results too. I'll explicitly look at the library terms to ensure that they are calculating the right mixed derivatives and push a fix.

Jacob-Stevens-Haas replied:
The only function called on `x` in FiniteDifference._differentiate is `np.full_like` and only attribute access is `x.shape`.  So yes, it's passed by reference, but it shouldn't be mutated.  If it does, that seems like an issue to add in itself.

znicolaou replied:
yes, FiniteDifference does not modify x(=derivs), but the assignment `derivs=FiniteDifference(derives)` will then modify x, so that x itself keeps accumulating all the previously calculated derivatives. So we need to make a copy of x each time we want to calculate a mixed derivative

Jacob-Stevens-Haas replied:
Aaah I see why it needs to store the mutation across `for axis in range(self.ind_range)`.  `derivs = x` should work (it does, just verified that it isn't mutated) but now that I understand the copy happens regardless, it's probably fine to make it explicit.

znicolaou replied:
Good! I don't remember when I added the .copy() it if I was reaching for straws debugging something, or if there was some case where it was necessary, but I think the way that I got it right was just to plot the derivatives computed by transform for a known input data and verify the mixed derivatives are correct. I'll do a quick check on the code before we finish the review to verify it does what its supposed to. Planning to take a final look at things this weekend, and draft the release notes.

znicolaou replied:
The update looks good to me, I ran this quick check:

```
import matplotlib.pyplot as plt
import numpy as np
import pysindy as ps

x=np.arange(128)/128
spatial_grid=np.zeros((128,128,2))
spatial_grid[:,:,0]=x[:,np.newaxis]
spatial_grid[:,:,1]=x[np.newaxis,:]
lib=pde_lib = ps.PDELibrary(derivative_order=2, spatial_grid=spatial_grid)
x=(np.sin(2*np.pi*np.arange(128)[:,np.newaxis]/128)*np.sin(4*np.pi*np.arange(128)[np.newaxis,:]/128))[:,:,np.newaxis]
lib.fit(x)
mat=lib.transform(x)

for i in range(mat.shape[-1]):
    plt.imshow(mat[:,:,i])
    plt.colorbar(label=lib.get_feature_names()[i])
    plt.show()
```
I'll mark this resolved.

Jacob-Stevens-Haas reviewed (commented):
Many of these can be turned into GH Issues/project todos.

Jacob-Stevens-Haas commented:
Hey @znicolaou , I finished reading through the PR and there were two questions I had for you on code that came from #195.  Also, I didn't reread feature_libraries code too closely.  When you review, can you give that part more focus than I did?

Jacob-Stevens-Haas commented on pysindy/utils/base.py:
```diff
@@ -15,11 +14,27 @@
 T_DEFAULT = object()
 
 
+def flatten_2d_tall(x):
+    return x.reshape(x.size // x.shape[-1], x.shape[-1])
+
+
 def validate_input(x, t=T_DEFAULT):
+    """Forces input data to have compatible dimensions, if possible.
+
+    Args:
+        x: array of input data (measured coordinates across time)
+        t: time values for measurements.
+
+    Returns:
+        x as 2D array, with time dimension on first axis and coordiante
+        index on second axis.
+    """
     if not isinstance(x, np.ndarray):
         raise ValueError("x must be array-like")
     elif x.ndim == 1:
         x = x.reshape(-1, 1)
+    x_new = flatten_2d_tall(x)
+
```
Defer till later PR in #206

Jacob-Stevens-Haas commented on pysindy/utils/base.py:
```diff
@@ -38,181 +53,87 @@ def validate_input(x, t=T_DEFAULT):
+    y_non_sample_axes = tuple(ax for ax in range(y.ndim) if ax != y.ax_sample)
+    x_good_samples = (~np.isnan(x)).any(axis=x_non_sample_axes)
+    y_good_samples = (~np.isnan(y)).any(axis=y_non_sample_axes)
+    good_sample_ind = np.nonzero(x_good_samples & y_good_samples)[0]
```
Nope, nvm, that's not how np.nonzero works.

znicolaou commented:
Thanks for the review @Jacob-Stevens-Haas! I will plan to take a look in the few days as well, and respond to your questions.

Jacob-Stevens-Haas commented on pysindy/utils/axes.py:
```diff
@@ -0,0 +1,238 @@
+                "given.  This occurs when reshaping data rather than creating "
+                "a new AxesArray with determined axes.",
+                type("AxesWarning", (PendingDeprecationWarning,), {}),
+            )
```
#220

Jacob-Stevens-Haas commented on pysindy/utils/axes.py:
```diff
@@ -0,0 +1,238 @@
+            warnings.warn(
+                "Concatenating >1 AxesArray with incompatible axes",
+                PendingDeprecationWarning,
+            )
```
Yes - is an error, fixed code that raised it.

znicolaou commented on test/feature_library/test_feature_library.py:
```diff
@@ -225,7 +232,9 @@ def test_sindypi_library_bad_params(params):
         pytest.lazy_fixture("data_generalized_library"),
         pytest.lazy_fixture("data_ode_library"),
-        pytest.lazy_fixture("data_pde_library"),
+        # pytest.lazy_fixture("data_pde_library"),
+        # we should not use this library for lorenz data,
+        # since spatiotemporal grid determines derivative order
```
"order" was not the right word to use here, but the issue is that the PDELibrary inspects the shape of the spatiotemporal grid (`self.ind_range = spatiotemporal_grid.ndim - 1`) to determine which derivatives to include in the library. This old test incorrectly supplied a library that used a 1d spatiotemporal grid to use on the Lorenz ODE data (a 0d system).  The old test passed before we changed some reshaping, but it would have produced a nonsensical fit (the data was forced into the right shape in a nonsensical way). When we changed the reshaping procedure, it began to throw an error, which just pointed out that the test didn't make sense in the first place. I could have just deleted the pytest.lazy_fixture and not included comments, but that may leave some case untested. A new test_fit_transform_pde could also be made to use pde data to test fit_transform with the pde and weakpdelibrary.

znicolaou replied:
PS: this points out that fit should really check the shapes of the data to ensure it is compatibile with the spatial/spatiotemporal grid, and raise an interpretable error if it isn't. Eventually I would like to allow different spatiotemporal grids to each trajectory in the multiple_trajectory case, which will probably require the spatial grid to be passed not at the initialization but later in the fit. But I'm planning to open a new pull request for this in the future.

Jacob-Stevens-Haas replied:
Sounds good - so should we add remove the comment, & add an issue to move stgrid initialization to `fit()`?  We don't need to solve this during the PR, I just want to make sure everything noticed during the PR ends up tracked.

znicolaou replied:
This sounds good to me. I think the feature_library.fit can't take additional arguments other than x and y, so what makes sense to me is that model.fit includes optional grid options, which, if present, calls a new feature_library.set_grid function to set the grid. Then, when feature_library.fit is called, it checks to see if the grid is None (then, use a default grid) or if it is not a compatible shape as the input data (then, raise an error).

znicolaou replied:
I added an issues #222 for the future. Feel free to push a change removing the test comments, and we can resolve this conversation. I will continue reviewing in the meanwhile.

Jacob-Stevens-Haas replied:
done :heavy_check_mark:

znicolaou commented on pysindy/utils/axes.py:
```diff
@@ -0,0 +1,187 @@
+    return axes
+
+
+def ax_time_to_ax_sample(x: AxesArray) -> AxesArray:
```
Is this function actually used for anything? It is applied in some but not all feature library transforms and in the SINDy score/predict. But the SampleConcatter implements a more general version of converting axes other axes to samples (which is why it doesn't matter that some feature libraries don't call ax_time_to_samples). I'd probably try to remove this, let feature libraries keep the axes unchanged, and use the SampleConcatter everywhere instead. But this is not pressing, so could be deferred.

Jacob-Stevens-Haas replied:
I'm happy to remove it - the function was created at a time when `drop_nan_samples` and `drop_random_samples` were called before the feature libraries `transform()`, and therefore before samples were concatted.  Now it probably can be removed.

Jacob-Stevens-Haas replied:
Removed in new commit

znicolaou commented on pysindy/feature_library/generalized_library.py:
```diff
@@ -232,7 +244,25 @@ def fit(self, x, y=None):
         return self
 
-    def transform(self, x):
+    def has_type(self, libtype: Type, exclusively=False) -> bool:
```
This is not used yet, but implemented for future purposes? Should the `ConcatLibrary` and `TensorLibrary` also implement it, and should included in `BaseFeatureLibrary`? Also, what if the self.libraries_ contain a `GeneralizedLibrary`, `TensorLibrary`, or `ConcatLibrary`? Should has_type recursively check the `has_type` of `self.libraries_`? It would be easier if we never need to check if the library contains any particular type and we could remove this function...

Jacob-Stevens-Haas replied:
Also a vestige of refactoring, when pysindy.py had several nested loops and conditionals to check whether to treat the data like it went to a `WeakPDELibrary`.

Since there's no current use or test for the function, I'm happy to delete it rather than try to improve it.

Jacob-Stevens-Haas replied:
removed in new commit.

znicolaou commented on pysindy/feature_library/polynomial_library.py:
```diff
@@ -222,66 +226,81 @@ def transform(self, x):
-        x = check_array(x, order="F", dtype=FLOAT_DTYPES, accept_sparse=("csr", "csc"))
+        xp_full = []
+        for x in x_full:
+            if sparse.issparse(x) and x.format not in ["csr", "csc"]:
```
The `PolynomialLibrary` is the only library that supports the sparse input data currently. I don't understand the use case for it--what sort of time series data is expected to be sparse? If there is a usage, should we make an issue to implement it more generally?

Jacob-Stevens-Haas replied:
The `PolynomialLibrary` is the only library that _optimizes_ for sparse input.  Sparse input may also work in other libraries via duck typing.

I think Kathleen added these lines when I ran git blame.  I had the same question.  It was just easier for me to build backwards compatibility rather than try to find out the reason.

I'll create an issue and reach out to Kathleen to see why she added them.

znicolaou replied:
Cool, makes sense! Obviously, nothing required for this PR, but just an observation.

znicolaou commented on pysindy/feature_library/base.py:
```diff
@@ -99,7 +159,7 @@ def _ensemble(self, xp):
             inds = range(self.n_output_features_)
             inds = np.delete(inds, self.ensemble_indices)
-            return xp[:, inds]
+            return [x[..., inds] for x in xp]
```
The ellipses `...` are used in many libraries to slice only along the `ax_coords` axis. For future usage where the axes aren't necessarily ordered with `ax_coord` at the end, it would be nice to include a utility function to slice along the `ax_coord`.

Jacob-Stevens-Haas replied:
`x.take(inds, axis=x.ax_coord)`

znicolaou replied:
Also, not sure this will work for assignment of slices along the `ax_coord` axis, like `spatiotemporal_grid[..., -1] = temporal_grid`. Would be easy to implement with a slice object, like 
```
s=[slice(None)]*spatiotemporal_grid.ndim
s[spatiotemporal_grid.ax_coord]=-1
spatiotemporal_grid[tuple(s)] = -1
```
Not sure whether a utility function in AxesArray would simplify code like this or not, but maybe.

Jacob-Stevens-Haas replied:
I haven't used it, it looks like `np.put_along_axis` seems to be the counterpart to `np.take`

znicolaou replied:
Cool, good to know, thanks!

znicolaou reviewed (approved):
I think I'm satisfied with the current state of this branch now, and am ready to draft some release notes, which I'll start thinking about this weekend. @Jacob-Stevens-Haas I left a few comments with questions and suggestions for the future, which we may want to include as issues for the future or make some small adjustments still before the merge. Let me know what you think.

Jacob-Stevens-Haas commented:
Merged!  I added a `1.x` branch at the old position of `master` just in case something goes wrong.

znicolaou commented:
Awesome, looks great! We just need to draft up the release notes to make a new release soon now.
